### PR TITLE
generic workload-controller sets status on preconditions  

### PR DESCRIPTION
**Pull request by p0lyn0mial**
this PR changes the generic `workload` controller used by `oas-o` and `cao` to set an operator's status to `degraded`, `not available` and `not progressing` when preconditions are not met.

Errors reported from `PreconditionFulfilled() (bool, error)` method will be reported in the `Message` message field, if empty then this will be set to `the operator haven't specified what preconditions are missing`

### CODE REVIEW

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 1]
```diff
@@ -273,6 +273,34 @@ func (c *Controller) updateOperatorStatus(workload *appsv1.Deployment, operatorC
 		Status: operatorv1.ConditionFalse,
 	}
 
+	if !preconditionsReady {
+		message := "the operator haven't specified what preconditions are missing"
```
Comment (stlaz) [inline pkg/operator/apiserver/controller/workload/workload.go:277]: nit: either "hasn't specified which" but I think "did not specify which" is more correct

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 2]
```diff
@@ -273,6 +273,34 @@ func (c *Controller) updateOperatorStatus(workload *appsv1.Deployment, operatorC
...
+		for _, err := range errs {
+			message = message + err.Error() + "\n"
+		}
```
Comment (stlaz) [inline pkg/operator/apiserver/controller/workload/workload.go:280]: message = strings.Join(errs, "\n")

**Review by stlaz — COMMENTED** [review]
nits

File: pkg/operator/apiserver/controller/workload/workload.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/workload/workload.go:280]: string.Joins accepts `[]string`, won't accept `errs` which is `error`

`string.Join(elems []string, sep string) string`

Comment (stlaz) [reply pkg/operator/apiserver/controller/workload/workload.go:280]: :facepalm: I did not realize we're dealing with errors, sorry, my bad

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 3]
```diff
@@ -117,12 +117,12 @@ func (c *Controller) sync() error {
 	}
 
 	if fulfilled, err := c.delegate.PreconditionFulfilled(); !fulfilled || err != nil {
```
Comment (sttts) [inline pkg/operator/apiserver/controller/workload/workload.go:119]: what are preconditions in practice in (all) our operators?

Comment (sttts) [reply pkg/operator/apiserver/controller/workload/workload.go:119]: We are changing what a precondition is here. This should go into the go doc in line 44.

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 4]
```diff
@@ -273,6 +273,34 @@ func (c *Controller) updateOperatorStatus(workload *appsv1.Deployment, operatorC
 		Status: operatorv1.ConditionFalse,
 	}
 
+	if !preconditionsReady {
+		message := "the operator didn't specify what preconditions are missing"
+		for _, err := range errs {
+			message = message + err.Error() + "\n"
```
Comment (sttts) [inline pkg/operator/apiserver/controller/workload/workload.go:279]: space missing for the first appended error.

Comment (sttts) [reply pkg/operator/apiserver/controller/workload/workload.go:279]: should we replace the placeholder message if errs is given ?

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 5]
```diff
@@ -273,6 +273,34 @@ func (c *Controller) updateOperatorStatus(workload *appsv1.Deployment, operatorC
...
+
+		deploymentDegradedCondition.Status = operatorv1.ConditionTrue
+		deploymentDegradedCondition.Reason = "PreconditionNotReady"
+		deploymentDegradedCondition.Message = message
+
+		deploymentAvailableCondition.Status = operatorv1.ConditionFalse
+		deploymentAvailableCondition.Reason = "PreconditionNotReady"
+
+		deploymentProgressingCondition.Status = operatorv1.ConditionFalse
+		deploymentProgressingCondition.Reason = "PreconditionNotReady"
```
Comment (sttts) [inline pkg/operator/apiserver/controller/workload/workload.go:292]: how is progressing defined? Isn't waiting for preconditions a variant of progressing?

File: pkg/operator/apiserver/controller/workload/workload.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/workload/workload.go:119]: for OAS-O and CAO preconditions are:

1. wait for the config
2. make sure that mandatory arguments passed to the aggregated APIs are present (i.e. etcd URLs) (this is something I'm going to add)

File: pkg/operator/apiserver/controller/workload/workload.go (hunk 4 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/workload/workload.go:279]: yes we should

File: pkg/operator/apiserver/controller/workload/workload.go (hunk 5 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/workload/workload.go:292]:
```
// Progressing indicates that the operator is trying to transition the operand to a different state
OperatorStatusTypeProgressing = "Progressing"
```

when preconditions are not ready we are not trying to transition the operand (no progress).

**Comment by p0lyn0mial** [comment]
comments applied, PTAL.

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 6]
```diff
@@ -273,6 +277,37 @@ func (c *Controller) updateOperatorStatus(workload *appsv1.Deployment, operatorC
...
+			message = message + err.Error() + "\n"
+		}
+		if len(message) == 0 {
+			message = "the operator didn't specify what preconditions are missing"
+		}
+
+		// we are degraded, not available and we are not progressing
+
+		deploymentDegradedCondition.Status = operatorv1.ConditionTrue
+		deploymentDegradedCondition.Reason = "PreconditionNotReady"
```
Comment (sttts) [inline pkg/operator/apiserver/controller/workload/workload.go:292]: ready is fixed term in kube. PreconditionNotFulfilled?

Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/workload/workload.go]: done.

**Pull request merged**